roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3788167&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3788167&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on release-21.1 @ [9c32f4116126c15fe5ac26ba62e4b4928b5ae371](https://github.com/cockroachdb/cockroach/commits/9c32f4116126c15fe5ac26ba62e4b4928b5ae371):


```
The test failed on branch=release-21.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:211,sqlsmith.go:240,test_runner.go:733: error: pq: internal error: estimated row count must be non-zero
		stmt:
		SELECT
			tab_5321.tableoid AS col_10902
		FROM
			defaultdb.public.table3@[0] AS tab_5321,
			defaultdb.public.table1 AS tab_5322
			JOIN defaultdb.public.table1 AS tab_5323 ON (tab_5322.col1_0) = (tab_5323.col1_0),
			defaultdb.public.table1@[0] AS tab_5324,
			(SELECT B'1010010101101011100000000111111010010100' AS col_10900 LIMIT 17:::INT8) AS tab_5325 (col_10901)
		WHERE
			st_coveredby(tab_5322.col1_0::GEOMETRY, '0107000000040000000103000000010000000D00000020D9A46769B9B641C83042A18C49DA4190833A756B78CEC1E8587C557028D44198B8C1569370D4C1D19FD2F4AE4702C230EEA16FFF0FDB4148EE7AD2B68CDFC1042ACA1F211AEA4160AF1DCD022DFAC1CCAED756509EF741D8690725DCE5D3C138101C6E5AA1F241487B97B4AD03D641EEF479B34C35FA41D81A8E62E7EAF1412AE39A5100710142BAA11CBEB54B0142443C3B32F393F741F45613B2BC0D0242001A2299B174B9C150B9AE07FBEAF04138C4509D1A74FFC1C292FB2F02D8004220D9A46769B9B641C83042A18C49DA410105000000050000000102000000030000005894D84BF379D84132E5EF8A272FE8C148400DA1BE76FB41CA0BE9C2B4C8E6C1F2C7157232F9F141E0CC1F227DC5CB410102000000030000006E832F5ABA91024284FDBDA6ED35EEC1E4EB2B75456D00428EF60C2FAFE8F341C82E4D066BAEF441667AFA81781901420102000000020000005EBA9D9A8A5AF6410A93A84AF91EFFC1529EBF9C12C901424E0036D6CF06FA41010200000005000000702FDC62421CCBC17426EDE9245FEFC1B49D7FFF0D3AEB410CAD8950394800C2A88FC8012129F641EED7B2E5BDD3F6C14E85A05012BFFA4178F820A72BD1FE41A2242305BDD3F5C1E698AA703E3FF641010200000007000000BCB616A92AA1F9C1C73A4615D83102C23CE7AE823107DAC17C53D0319494F7C1E8781F07D882DC41FD2C58779EF7FEC1947F060F4207EC41587E55C38F6300C27C95DD9BED6DE941D47A8EE69B40FDC1DEBF2F1A150A024288AFA4E97769FEC1A82B15DAC7A0DD4100050567D4CDC5C101060000000A00000001030000000100000005000000E25F8D266921F541944FA699CF42EC4158C50F669B3D0242F8836B2ED77DD2C1F0EA873394120142080CCA7A7BF00142D4EBEC4C7F4EEDC140F68F85FA5DF541E25F8D266921F541944FA699CF42EC4101030000000100000009000000B83F6264A540D341BEAD39B72739E8C1DC341FCEFC85E841EA5087372741E5C1F83A492B6745E44172BA8333A86AF4C1C8EF8D2B5A0BFD41078E8779685D01C2D87B1E74E3FAF541409967275DA5B9C112C7BB03228A00428852D87C7572F341942B9EF85765F7419A306D191D5FF3416CB2A2D84E88F241008D36F1F0E3B3C1B83F6264A540D341BEAD39B72739E8C10103000000010000000C000000E4BFAC4ACFBAEFC14410D93C3748D5C13077EC539BD7FBC1BE76B9A074BBE6C1C66A29707B34E2C1A27B11111690FCC1CC8FA6E55423E14154B52A375BD2FDC13459A457E382F9414DFC9120211B02C266FF6F3D3DB3F641DC88D92DB9BAEAC1B4B6321F0C5BFE41048D89778696EB41127FA55471A8E5C1766FBFBA7539F5414B66E46074B201C29C2ACE2D55E70042200B092CC1F4DBC1C0AC2BB7F5ABA4C1D953A02DAC1C01C2F8E3F4896B29F541E4BFAC4ACFBAEFC14410D93C3748D5C101030000000100000004000000E2606552F6F5F7C142839146D69C02C2DE6D64BECB1A02425AFCB50FC200F1417698DF46836EF741F8582D7360780142E2606552F6F5F7C142839146D69C02C20103000000010000000C0000002FAAF9B974AE00C23CE4C4B63E09FCC1CA4AEBB8373800C2180ECDEE32B600C2C0B9DAE6F6A2CBC1987B04FD7E7BF2C1746D4CA18BCEFA419EFF94B15CFDF8C1F413BAC9E036F14138C61633E371EAC1E48D99CC3AE4FE41B6F1F2D1E98BE6C15A40315CE92D014294662AE1A1CBEF41A6E89F2595D90142A8F3673A2B8CFE4168993C68360EF841047C17F8D927FF417CA88DF3FBB5DDC1E0A9CD9C5DEAB541E2C97751E00CF4C1503F998EDF69D1C12FAAF9B974AE00C23CE4C4B63E09FCC101030000000100000008000000A28EC7DBA9F8FDC1229E5C0FEB03E5C13F1A8558C681F0C1FE747F4DDC73F0C13831A202B9F2EF4180C89CB7D265F9C17E3F72EA98BB014268798928E706FC41B0C212E1AE5FC84168CCD25F7052FD4188E52DFD87DDD3C1FCFC1A85EF3302424CBFB6F00CEBE0C1EADFEC88E01F0042A28EC7DBA9F8FDC1229E5C0FEB03E5C10103000000010000000A000000C098E3BE7A6BA9C13210BD5E2B2FFFC1988E22E23788FC418405B68B85C200C2ACA1B7AC0D85EA41C0D2217294FFCFC1946E25D058F4FA416A5F9B5099F6F1C110D4BA37DC17FF4180D07220B679C2418860458627A1DB41FA834AC273710242C808A4A7FA3BD8C1C8E97D991ADCFB411836DDCC0DD5F0C15A3EC14176570042E74980996E4E00C2B6180F6EB186F741C098E3BE7A6BA9C13210BD5E2B2FFFC101030000000100000004000000F85D008B2B7CEBC164A79D181DDDE6C1E202C4E1D637FA41EC4ACB405042F041F08B4EA8C6FDE641F00BF7E65B66F941F85D008B2B7CEBC164A79D181DDDE6C10103000000010000000400000000E5E7931995AB4104EA7D4CBD74D5C170714B681089FE41B44FBF1C5D7FFE411AFD294C75DBE1C156884D193BD8F44100E5E7931995AB4104EA7D4CBD74D5C101030000000100000006000000CA71752ACA3DF7C1C413C339F382E341BC2CC9E86BA701C2EA001A5BAC69E0C170EF0F469F2AF341B8B55F4F3712E241EE65340CD64D01420CE8AB6B4514FE41A443258F5890DCC15AFD693A2DD80142CA71752ACA3DF7C1C413C339F382E3410101000000F5DFE5F174FC00C2AC8C86A5CF670042':::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			80:::INT8;
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=no-mutations
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -577,8 +577,10 @@ func (sb *statisticsBuilder) makeTableStatistics(tabID opt.TableID) *props.Stati
 							if invColStat, ok := stats.ColStats.Add(invCols); ok {
 								invColStat.Histogram = &props.Histogram{}
 								invColStat.Histogram.Init(sb.evalCtx, invCol, stat.Histogram())
-								// Set inverted entry counts from the histogram.
-								invColStat.DistinctCount = invColStat.Histogram.DistinctValuesCount()
+								// Set inverted entry counts from the histogram. Make sure the
+								// distinct count is at least 1, for the same reason as the row
+								// count above.
+								invColStat.DistinctCount = max(invColStat.Histogram.DistinctValuesCount(), 1)
 								// Inverted indexes don't have nulls.
 								invColStat.NullCount = 0
 							}
@@ -747,11 +749,13 @@ func (sb *statisticsBuilder) constrainScan(
 
 			inputStat, _ := sb.colStatFromInput(colSet, scan)
 			if inputHist := inputStat.Histogram; inputHist != nil {
-				// If we have a histogram, set the row count to its total,
-				// unfiltered count. This is needed because s.RowCount is
-				// currently the row count of the table, but should instead
-				// reflect the number of inverted index entries.
-				s.RowCount = inputHist.ValuesCount()
+				// If we have a histogram, set the row count to its total, unfiltered
+				// count. This is needed because s.RowCount is currently the row count
+				// of the table, but should instead reflect the number of inverted index
+				// entries. (Make sure the row count is at least 1. The stats may be
+				// stale, and we can end up with weird and inefficient plans if we
+				// estimate 0 rows.)
+				s.RowCount = max(inputHist.ValuesCount(), 1)
 				if colStat, ok := s.ColStats.Lookup(colSet); ok {
 					colStat.Histogram = inputHist.InvertedFilter(scan.InvertedConstraint)
 					histCols.Add(invertedConstrainedCol)

--- a/pkg/sql/opt/memo/testdata/stats/scan
+++ b/pkg/sql/opt/memo/testdata/stats/scan
@@ -2094,3 +2094,71 @@ index-join t
       ├── stats: [rows=111.111111, distinct(4)=11.1111111, null(4)=0]
       ├── key: (1)
       └── fd: (1)-->(4)
+
+# Regression test for #73106. Ensure we don't get the error "estimated row count
+# must be non-zero".
+exec-ddl
+CREATE TABLE tab (
+  geom GEOMETRY NOT NULL,
+  INVERTED INDEX (geom ASC)
+)
+----
+
+exec-ddl
+ALTER TABLE tab INJECT STATISTICS '[
+  {
+    "columns": ["geom"],
+    "created_at": "2000-01-01 00:00:00+00:00",
+    "distinct_count": 7673676,
+    "histo_buckets": [
+      {
+        "distinct_range": 0,
+        "num_eq": 0,
+        "num_range": 0,
+        "upper_bound": "\\\\x42fdffffffffffffffff01c1f3f5e7590b574e41f15b57b0682314"
+      }
+    ],
+    "histo_col_type": "BYTES",
+    "name": "__auto__",
+    "null_count": 0,
+    "row_count": 20000000
+  }
+]'
+----
+
+opt
+SELECT * FROM tab
+WHERE st_coveredby(geom, 'POINT(1 1)':::GEOMETRY)
+----
+select
+ ├── columns: geom:1(geometry!null)
+ ├── immutable
+ ├── stats: [rows=2222222.22, distinct(1)=2222222.22, null(1)=0]
+ ├── index-join tab
+ │    ├── columns: geom:1(geometry!null)
+ │    ├── stats: [rows=1]
+ │    └── inverted-filter
+ │         ├── columns: rowid:2(int!null)
+ │         ├── inverted expression: /5
+ │         │    ├── tight: false, unique: false
+ │         │    └── union spans
+ │         │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x10\x00\x00\x00\x00\x00\x00\x00"]
+ │         │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x01", "B\xfd\x12\x00\x00\x00\x00\x00\x00\x00")
+ │         │         └── ["B\xfd\x14\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x14\x00\x00\x00\x00\x00\x00\x00"]
+ │         ├── pre-filterer expression
+ │         │    └── st_covers('0101000000000000000000F03F000000000000F03F', geom:1) [type=bool]
+ │         ├── stats: [rows=1]
+ │         ├── key: (2)
+ │         └── scan tab@tab_geom_idx
+ │              ├── columns: rowid:2(int!null) geom_inverted_key:5(geometry!null)
+ │              ├── inverted constraint: /5/2
+ │              │    └── spans
+ │              │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x10\x00\x00\x00\x00\x00\x00\x00"]
+ │              │         ├── ["B\xfd\x10\x00\x00\x00\x00\x00\x00\x01", "B\xfd\x12\x00\x00\x00\x00\x00\x00\x00")
+ │              │         └── ["B\xfd\x14\x00\x00\x00\x00\x00\x00\x00", "B\xfd\x14\x00\x00\x00\x00\x00\x00\x00"]
+ │              ├── stats: [rows=1, distinct(2)=1, null(2)=0, distinct(5)=1, null(5)=0]
+ │              │   histogram(5)=
+ │              ├── key: (2)
+ │              └── fd: (2)-->(5)
+ └── filters
+      └── st_coveredby(geom:1, '0101000000000000000000F03F000000000000F03F') [type=bool, outer=(1), immutable, constraints=(/1: (/NULL - ])]

